---
title: "transana_tissuebias"
author: "Haiwang Yang"
date: "6/19/2018"
---

```{r setup, include=FALSE}
library("pheatmap")
library("RColorBrewer")
library("ggplot2")
library(Hmisc)
library(tximportData)
library(tximport)
library(rjson)
library(ggplot2)
library(DESeq2)
```

# basic variables, data, and look-up tables
```{r basic variables, data, and look-up tables}
base_path = "/Users/yangh13/transana/transana/"
dir <- paste0(base_path, "data/")

tissues_in_final_order = c( "wb", "go", "re", "ge", "tx", "dg", "hd", "ac")
sexed_tissues_in_final_order = do.call(paste, expand.grid(c("f","m"), tissues_in_final_order))
tissues_in_final_order_no_ge = c( "wb", "go", "re", "tx", "dg", "hd", "ac")
sexed_tissues_in_final_order_no_ge = do.call(paste, expand.grid(c("f","m"), tissues_in_final_order_no_ge))

species = "dmel" # use dmel w1118 as a test


samples <- read.table(file.path(dir, paste0(species, ".design")), header = TRUE)
files <- file.path(dir, "salmon", "FB", samples$sample_id, "quant.sf")
names(files) <- samples$sample_id
tx2gene = read.table(paste0(dir, species, ".tx2gene"), header=TRUE)
FBgn2FBtr = read.table(paste0(dir, species, ".FBgn2FBtr"), header=TRUE)
gene2ortho = read.table(paste0(dir, species, ".ortholog"), header=TRUE)
gene2symbol = read.table(paste0(dir, species, ".FBgn2symbol"), header=TRUE)
```


# basic funtions
```{r basic funtions}
first = function(str) {v = strsplit(as.character(str), "\\_")[[1]]; paste(v[1])}
second = function(str) {v = strsplit(as.character(str), "\\_")[[1]]; paste(v[2])}
third = function(str) {v = strsplit(as.character(str), "\\_")[[1]]; paste(v[3])}
fourth = function(str) {v = strsplit(as.character(str), "\\_")[[1]]; paste(v[4])}

get_tissueFull = function(tissue) {gsub("wb", "whole\norganism", gsub("re", "reproductive\ntract", gsub("hd", "head", gsub("dg", "viscera", gsub("tx", "thorax", gsub("ge", "terminalia",gsub("ac", "abdomen", gsub("go","gonad", tissue))))))))}

fetch_files_by_sexed_tissue = function(strain, sex, tissue, files){
  return(files[ grepl(paste0(strain, "_"), files) & grepl(paste0("_", sex, "_"), files) & grepl(paste0("_", tissue, "_"), files)])
}

get_FBgn_for_symbol = function(target_gene){
  target_FBgn = as.character(gene2symbol$FBgn[gene2symbol$symbol == target_gene])
  return(target_FBgn)
}

get_FBtrs_for_symbol = function(target_gene){
  target_FBgn = get_FBgn_for_symbol(target_gene)
  target_FBtrs = as.character(FBgn2FBtr$FBtr[FBgn2FBtr$FBgn == target_FBgn])
  return(target_FBtrs)
}

```



# transcript-level salmon-tximport-deseq2 analyses of tissue-biased expression
```{r get transcript-level salmon-tximport-deseq2 analyses of tissue-biased expression}
analyze_tissue_bias = function(strain, sex, tissue, level, alpha = 0.05){    
  part_files0 = fetch_files_by_sexed_tissue(strain, sex, "wb", files)
  part_files1 = fetch_files_by_sexed_tissue(strain, sex, tissue, files)
  part_files = c(part_files0, part_files1)
  
  part_samples = samples[samples$sample_id %in% names(part_files),]
  part_samples$tissue = factor(part_samples$tissue, levels = c(tissue, "wb"))
  
  if (level == "gene"){
    txi <- tximport(part_files, type = "salmon", tx2gene = tx2gene) # gene-level output
  }
  
  if (level == "transcript"){
    txi <- tximport(part_files, type = "salmon", txOut = TRUE) # transcript-level output
  }
  
  dds <- DESeqDataSetFromTximport(txi, part_samples, ~tissue)
  dds = DESeq(dds)
  
  res = results(dds) # log2fc is f/m, so Yp1-transcript has 13
  res = res[order(row.names(res)),]
  
  res_sig = res[res$padj<alpha & !is.na(res$padj),]
  res_sig_wb = res_sig[res_sig$log2FoldChange < 0,] # whole body of the same sex
  res_sig_tt = res_sig[res_sig$log2FoldChange > 0,] # this sexed tissue
  FBids_res_sig_wb = rownames(res_sig_wb)  
  FBids_res_sig_tt = rownames(res_sig_tt)
  
  FBids_all = rownames(res)
  FBids_all_minus_m = FBids_all[! FBids_all %in% FBids_res_sig_tt]
  FBids_unbiased = FBids_all_minus_m[! FBids_all_minus_m %in% FBids_res_sig_wb]
  df_bias_m = data.frame("FBid_ID"=as.character(unlist(FBids_res_sig_tt)),"bias"="tt")
  df_bias_unbias = data.frame("FBid_ID"=FBids_unbiased,"bias"="unbiased")
  df_bias_f = data.frame("FBid_ID"=as.character(unlist(FBids_res_sig_wb)),"bias"="wb")
  df_bias_total = rbind(df_bias_m,df_bias_unbias,df_bias_f)
  df_bias_total$FBid_ID = as.character(df_bias_total$FBid_ID)
  df_bias_total = df_bias_total[order(df_bias_total$FBid_ID),]
  df_bias_total = df_bias_total[order(df_bias_total$FBid_ID),]
  
  normCounts = counts(dds, normalized=TRUE)
  x = as.data.frame(log2(normCounts+1))
  x = x[order(row.names(x)),]
  
  x$MAplot_mean = as.numeric(apply(x, 1, mean))
  x$MAplot_log2FoldChange = res$log2FoldChange
  x$padj = res$padj
  x$bias = df_bias_total$bias
  
  x$col = rgb(0,0,0,0.3)
  x$col[x$bias == "tt"] = rgb(0,1,0,0.6)
  x$col[x$bias == "wb"] = rgb(1,0,1,0.6)
  return(x)
}


transana_tb = list()
for (strain in c("w1118", "oreR", species_in_final_order[2:8], "dgriG1")){
  species = strain
  version = "YO"
  tissues = tissues_in_final_order_no_ge
  if (strain %in% c("w1118", "oreR")){
    species = "dmel"
    tissues = tissues_in_final_order
    version = "FB"
  }
  if (strain %in% c("dgriG1")){
    species = "dgri"
    tissues = tissues_in_final_order
  }
  samples <- read.table(file.path(dir, paste0(species, ".design")), header = TRUE)
  files <- file.path(dir, "salmon", version, samples$sample_id, "quant.sf")
  names(files) <- samples$sample_id
  tx2gene = read.table(paste0(dir, species, ".tx2gene"), header=TRUE)
  gene2ortho = read.table(paste0(dir, species, ".ortholog.txt"), header=TRUE)

  for (tissue in tissues[2:length(tissues)]){
    for (sex in c("f", "m")){
      x = analyze_tissue_bias(strain, sex, tissue, "transcript", 0.05)
      transana_tb[[paste0(strain, "_", sex, "_", tissue)]] = x
    }
  }
}

# save(transana_tb,file=paste0(base_path, "output/transana_tb.rtable"))
load(paste0(base_path, "output/transana_tb.rtable"))
```

# compare gene-level and transcript-level tests for two well-known genes Yp1 and fne
```{r compare gene-level and transcript-level tests for two well-known genes Yp1 and fne}
strain = "w1118"
sex = "m"
tissue = "hd"

g = analyze_tissue_bias(strain, sex,  tissue, "gene", 0.05)
t = analyze_tissue_bias(strain, sex, tissue, "transcript", 0.05)

g[get_FBgn_for_symbol("Yp1"),]

t[get_FBtrs_for_symbol("Yp1"),]


g[get_FBgn_for_symbol("fne"),]

t[get_FBtrs_for_symbol("fne"),]

```


# Isoform-level MAplots
```{r print isoform-level MAplots}

for (strain_sex_tissue in names(transana)){
  x = transana[[strain_sex_tissue]]
  
  png(paste0(base_path, "/output/", strain_sex_tissue, ".transcript.MAplot.png"),width=2000,height=2000,res=600)
  par(mar=c(1,1,1,1))
  x$MAplot_log2FoldChange[x$MAplot_log2FoldChange > 2] = 2
  x$MAplot_log2FoldChange[x$MAplot_log2FoldChange < -2] = -2
  x$MAplot_log2FoldChange[is.na(x$MAplot_log2FoldChange)] = 0
  x$MAplot_log2FoldChange = -x$MAplot_log2FoldChange
  plot(x$MAplot_mean, x$MAplot_log2FoldChange, col=x$col, pch = 19, xlim=c(0,18), cex.axis=0.01, xlab="", ylab="")
  dev.off()
}

```
  


```{r barplot summary for transcript-level tissuebias-MAplots}
df = data.frame()
for (strain in c("w1118", species_in_final_order[2:8], "dgriG1")){
  species = strain
  tissues = tissues_in_final_order_no_ge
  if (strain %in% c("w1118", "oreR")){
    species = "dmel"
    tissues = tissues_in_final_order
  }
  if (strain %in% c("dgriG1")){
    species = "dgri"
    tissues = tissues_in_final_order
  }
  
  for (tissue in tissues[2:length(tissues)]){
    for (sex in c("f", "m")){
      tb0 = table(transana_tb[[paste0(strain, "_", sex, "_", tissue)]]$bias)
      cat(strain, " ", sex, " ", tissue," ", tb0,"\n")
      df0 = as.data.frame(tb0)
      colnames(df0) = c("tissuebias", "num")
      df0$species = species
      df0$sex = sex
      df0$tissue = get_tissueFull(tissue)
      df = rbind(df, df0)
    }
  }
}


df = df[,c("species", "sex", "tissue", "tissuebias", "num")]

df$sst = as.character(interaction(df$species, df$sex, df$tissue)) # sst = species sex tissue
df$sst = factor(df$sst, levels = unique(df$sst))

pdf(paste0(base_path, "output/transcript.level.tb.bar.summary.pdf"),width=10,height=10,useDingbats=FALSE)
ggplot(df, aes(x = sst, fill= tissuebias, group= tissuebias, y= num)) + geom_bar(stat = "identity") + scale_fill_manual(values = c("green", "grey", "purple")) + theme_bw() + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + ylab("number of genes")
dev.off()

```



```{r correlation of transcript-level tissue-bias between w1118 and oreR}
df = data.frame()
for (tissue in tissues[2:length(tissues)]){
  for (sex in c("f", "m")){
    A = table(transana_tb[[paste0("w1118_", sex, "_", tissue)]]$bias)
    B = table(transana_tb[[paste0("oreR_", sex, "_", tissue)]]$bias)
    
    df0 = data.frame("w1118" = as.numeric(A["tt"] + A["wb"]), "oreR" = as.numeric(B["tt"] + B["wb"]))
    df = rbind(df, df0)
  }
}
row.names(df) = tissues

pdf(paste0(base_path, "output/transcript.level.tb.scatterplot.w1118.oreR.pdf"),width=5,height=5,useDingbats=FALSE)
plot(df$w1118, df$oreR, xlab="number of transcripts with tissue-biased expression in w1118", ylab="number of transcripts with tissue-biased expression in oreR")
text(8000, 8000, paste0("Pearson's r = ", round(cor(df$w1118, df$oreR), 3)))
dev.off()

```


# expression of fne isoforms in boxplot
```{r expression fne two isoforms in boxplot}
target_gene = "fne"

target_FBtrs = get_FBtrs_for_symbol(target_gene)

target_nrc = data.frame(row.names = target_FBtrs)
target_lfc = data.frame(row.names = target_FBtrs)
target_bias = data.frame(row.names = target_FBtrs)
for (strain_tissue in names(transana)){
  x = transana[[strain_tissue]]
  
  target_nrc = cbind(target_nrc, x[target_FBtrs,1:8])
  
  tmp_target_lfc = data.frame(row.names = target_FBtrs, a = x[target_FBtrs,]$MAplot_log2FoldChange)
  colnames(tmp_target_lfc) = strain_tissue
  target_lfc = cbind(target_lfc, tmp_target_lfc)
  
  tmp_target_bias = data.frame(row.names = target_FBtrs, a = x[target_FBtrs,]$bias)
  colnames(tmp_target_bias) = strain_tissue
  target_bias = cbind(target_bias, tmp_target_bias)
}

target_nrc = data.frame(t(target_nrc))

target_nrc$strain = sapply(row.names(target_nrc), first)
target_nrc$sex = sapply(row.names(target_nrc), second)
target_nrc$tissue = sapply(row.names(target_nrc), third)
target_nrc$rep = sapply(row.names(target_nrc), fourth)
target_nrc$sex_tissue = interaction(target_nrc$sex, target_nrc$tissue, sep = " ")
target_nrc$sex_tissue = factor(target_nrc$sex_tissue, levels = sexed_tissues_in_final_order)
target_nrc$fulltissue = as.character(sapply(target_nrc$tissue, get_tissueFull))
target_nrc$fulltissue = factor(target_nrc$fulltissue, levels = unique(target_nrc$fulltissue))

target_nrc$sex = as.factor(target_nrc$sex)

target_nrc$sex = factor(target_nrc$sex, levels= c("m","f"))

pdf(paste0(base_path, "/output/fne.two.isoform.nrc.boxplot.pdf"),width=6,height=5)
par(mfrow=c(1,2))
ggplot(data = target_nrc, aes(x = target_nrc$fulltissue, y = FBtr0306738, fill = target_nrc$sex)) + geom_boxplot(outlier.size=-1) + theme_bw() + scale_fill_manual(values = c("blue", "red")) + ylab("log2 (normalized read count + 1) of isoform FBtr0306738")
ggplot(data = target_nrc, aes(x = target_nrc$fulltissue, y = FBtr0306739, fill = target_nrc$sex)) + geom_boxplot(outlier.size=-1) + theme_bw() + scale_fill_manual(values = c("blue", "red")) + ylab("log2 (normalized read count + 1) of isoform FBtr0306739")
dev.off()
```






# compare and summarize gene and transcript level salmon-tximport-deseq2 analyses
```{r compare gene and transcript level salmon-tximport-deseq2 analyses of sex-biased expression}
c = data.frame(row.names=FBgn2FBtr$FBtr, FBgn = FBgn2FBtr$FBgn)

alltab = data.frame(row.names = c("FALSE.FALSE.FALSE.FALSE", "TRUE.FALSE.FALSE.FALSE", "FALSE.TRUE.FALSE.FALSE","TRUE.TRUE.FALSE.FALSE", "FALSE.FALSE.TRUE.FALSE", "TRUE.FALSE.TRUE.FALSE", "FALSE.TRUE.TRUE.FALSE", "TRUE.TRUE.TRUE.FALSE", "FALSE.FALSE.FALSE.TRUE", "TRUE.FALSE.FALSE.TRUE", "FALSE.TRUE.FALSE.TRUE", "TRUE.TRUE.FALSE.TRUE", "FALSE.FALSE.TRUE.TRUE", "TRUE.FALSE.TRUE.TRUE", "FALSE.TRUE.TRUE.TRUE", "TRUE.TRUE.TRUE.TRUE" ))

for (strain in c("w1118")){
  for (tissue in tissues_in_final_order){
    g = analyze_sex_bias(strain, tissue, "gene", 0.05)
    
    t = analyze_sex_bias(strain, tissue, "transcript", 0.05)
    
    target_FBtrs = get_FBtrs_for_symbol(target_gene)
    
    tt = merge(t, c, by="row.names")
    
    gs = as.data.frame.matrix(table(row.names(g), g$bias))
    ts = as.data.frame.matrix(table(tt$FBgn, tt$bias))
    ss = merge(gs, ts, by="row.names")
    row.names(ss) = ss$Row.names
    ss = ss[,-1]
    
    sss = ss>0

    # changing of sex-bias    
    tab = table(interaction(sss[,'female.x'], sss[,'male.x'], sss[,'female.y'], sss[,'male.y']))
    tab = data.frame(row.names = row.names(tab), a = as.numeric(tab))
    colnames(tab) = paste0(strain, "_", tissue)
    
    alltab = cbind(alltab, tab)
  }
}
  
```



